pbj-core/pbj-runtime/src/main/java/com/hedera/hashgraph/pbj/runtime/io/ReadOnlyDataBuffer.java

```diff
@@ -0,0 +1,230 @@
+package com.hedera.hashgraph.pbj.runtime.io;
+
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A Buffer backed by a ByteBuffer that implements {@code DataInput} and {@code DataOutput}.
+ */
+public sealed class ReadOnlyDataBuffer implements DataInput permits DataBuffer, ReadOnlyOffHeapDataBuffer {
+
+    /** ByteBuffer used as backing buffer for this DataBuffer */
+    protected ByteBuffer buffer;
+
+    /**
+     * Wrap an existing allocated ByteBuffer into a DataBuffer. No copy is made.
+     *
+     * @param buffer the ByteBuffer to wrap
+     */
+    protected ReadOnlyDataBuffer(ByteBuffer buffer) {
+        this.buffer = buffer;
+    }
+
+    /**
+     * Allocate new on Java heap Data buffer
+     *
+     * @param size size of new buffer in bytes
+     */
+    protected ReadOnlyDataBuffer(int size) {
+        this.buffer = ByteBuffer.allocate(size);
+    }
+
+    // ================================================================================================================
+    // Static Builder Methods
+
+    /**
+     * Wrap an existing allocated ByteBuffer into a ReadOnlyDataBuffer. No copy is made.
+     *
+     * @param buffer the ByteBuffer to wrap
+     * @return new DataBuffer using {@code buffer} as its data buffer
+     */
+    public static ReadOnlyDataBuffer wrap(ByteBuffer buffer) {
+        return buffer.isDirect() ? new ReadOnlyOffHeapDataBuffer(buffer) : new ReadOnlyDataBuffer(buffer);
+    }
+
+    // ================================================================================================================
+    // DataBuffer Methods
+
+    /**
+     * Reset position to origin and limit to capacity, allowing this buffer to be read or written again
+     */
+    public void reset() {
+        buffer.clear();
+    }
+
+    /**
+     * Get the capacity in bytes that can be stored in this buffer
+     *
+     * @return capacity in bytes
+     */
+    public int getCapacity() {
+        return buffer.capacity();
+    }
+
+    /**
+     * Create a new sub data buffer over a subsection of this buffer. Data is shared and not copied, so any changes to
+     * the contents of this buffer will be reflected in the sub-buffer. The sub buffer is always read only.
+     *
+     * @param length The length in bytes of this buffer starting at current position to be in sub buffer
+     * @return new read only data buffer representing a subsection of this buffers data
+     * @throws BufferUnderflowException If length is more than remaining bytes
+     */
+    public ReadOnlyDataBuffer readDataBuffer(int length) {
```

Comment:
I wonder if we should call the method slice

```suggestion
    public ReadOnlyDataBuffer slice(int length) {
```